plugins/field-dependent-dropdown/src/field_dependent_dropdown.ts

```diff
@@ -0,0 +1,266 @@
+   *    field's value.
+   * @param config A map of general options used to configure the field, such as
+   *    a tooltip.
+   */
+  constructor(
+      parentName: string,
+      optionMapping: ChildOptionMapping,
+      validator?: Blockly.FieldValidator,
+      config?: FieldConfig) {
+    // A menu generator needs to be passed to the super constructor, but it
+    // needs to be able to reference data that hasn't been populated yet. We're
+    // not allowed to refer to "this" in this constructor before calling
+    // "super", so let's make separate structure to hold data relevant to the
+    // menu generator and populate that later.
+    const dependencyData: DependencyData = {};
+
+    // A menu option generator function for this child field that reads the
+    // derived options in the dependency data if available.
+    const menuGenerator: Blockly.MenuGeneratorFunction = () => {
+      // If derivedOptions has been initialized, use that.
+      if (dependencyData.derivedOptions) {
+        return dependencyData.derivedOptions;
+      }
+
+      // Fall back on the options corresponding to the parent field's current
+      // value (which is fine when initializing but may be out of date when
+      // making changes since the parent field's validator function triggers
+      // this function before the parent field's value is updated).
+
+      // If the parent field exists, and its value is a key in the provided
+      // option mapping, use the corresponding options.
+      if (dependencyData.parentField) {
+        const options = optionMapping[dependencyData.parentField.getValue()];
+        if (options) {
+          return options;
+        }
+      }
+
+      // Fall back on basic default options.
+      return [['', '']];
+    };
+
+    super(menuGenerator, validator, config);
+    this.parentName = parentName;
+    this.optionMapping = optionMapping;
+    this.dependencyData = dependencyData;
+  }
+
+  /**
+   * Constructs a FieldDependentDropdown from a JSON arg object.
+   * @param options A JSON object providing "parentName" and "optionMapping".
+   * @returns The new field instance.
+   */
+  static fromJson(
+      options: FieldDependentDropdownFromJsonConfig): FieldDependentDropdown {
+    return new FieldDependentDropdown(
+        options['parentName'],
+        options['optionMapping'],
+        undefined,
+        options);
+  }
+
+  /**
+   * Attach this field to a block.
+   *
+   * @param block The block containing this field.
+   */
+  setSourceBlock(block: Blockly.Block) {
+    super.setSourceBlock(block);
+
+    const parentField: Blockly.Field<string> | null =
+        block.getField(this.parentName);
+
+    if (!parentField) {
+      throw new Error('Could not find a parent field with the name ' +
+          this.parentName + ' for the dependent dropdown.');
+    }
+
+    if (parentField.getValidator()) {
+      throw new Error('The parent field with the name ' +
```

Comment:
Instead of throwing an error, can you let the parent field have its own validator, and then call the old validator from the new one you set? Because especially if the parent field's value should be restricted to a specific set of values (that are supported in the dependent dropdown), it would be intuitive for that work to occur in the validator originally passed to the parent field.

As it is, you are allowing the value to be set to something not in the config options list (by returning undefined from the validator in that case, which allows the newValue to be set as is), so there is no way to actually restrict the value of the field, which you could imagine someone wanting to do.

Reply:
Done, although for timing reasons it only works if you use the full JavaScript API. If a user's extension adds a validator to the parent field, that will happen after the dependency has been initialized along with its built-in validator, in which case it doesn't know yet about the one the extension wants to add and the extension would just end up replacing the one set up by the dependency. 

I guess the user's extension could compose the validator functions together similar to how this plugin does it, if they know what they're doing, but it wouldn't be obvious. 

I'll add a description to the README about how to use the JavaScript API to add a validator to the parent.